### PR TITLE
implement cheat menu, add a few cheats from DSDA-Doom

### LINKED ISSUE #922: [Feature] bind cheats to keys
I'd love to be able to bind specific keys to activate a cheat in the keybindings menu as typing any cheat that starts with "id" causes me to move to the right after clicking D

### PR DESCRIPTION
**Pull request by rfomin**
* Add cheat menu from DSDA-Doom.

* Add FREEZE cheat from DSDA-Doom.

Fix #922

### CODE REVIEW

**Comment by rfomin** [comment]
woof0000

Perhaps we should remove or add some of the cheats. Also we are missing `idbeholdh` (mega health) and `idbeholdm` (mega armor).

**Comment by fabiangreffrath** [comment]
> Perhaps we should remove or add some of the cheats. 

I went on to remove some of the cheats a while back, but couldn't decide for the most obvious candidates.

> Also we are missing `idbeholdh` (mega health) and `idbeholdm` (mega armor).

Should be straightforward to add them.

**Comment by fabiangreffrath** [comment]
> I went on to remove some of the cheats a while back, but couldn't decide for the most obvious candidates.

Discussion in #800, PR in #802.

File: src/m_cheat.c [hunk 1]
```diff
@@ -1046,6 +1068,53 @@ boolean M_FindCheats(int key)
   return ret;
 }
 
+typedef struct {
+  int input;
+  const cheat_when_t when;
+  const cheatf_t func;
```
Comment (fabiangreffrath) [inline src/m_cheat.c:1074]: `cheatf_t` is a union. I think you will have to initialize it using the explicit element name, e.g. `.v = cheat_god` or `.i = cheat_pw`.

**Comment by rfomin** [comment]
> I went on to remove some of the cheats a while back, but couldn't decide for the most obvious candidates.

I mean add or remove from this menu. For example, do we need `idmypos` here? It's probably for speedrunners.

File: src/m_cheat.c (hunk 1 above)
Comment (rfomin) [reply src/m_cheat.c:1074]: Fixed in https://github.com/fabiangreffrath/woof/pull/1001/commits/05fb48983d7ae16ba1b41557b237e40a497d28c3

Comment (fabiangreffrath) [reply src/m_cheat.c:1074]: Even better! 👍

**Comment by fabiangreffrath** [comment]
> Should be straightforward to add them.

Hm, on the other hand, these two just *give a specific item*. Aren't IDBEHOLD cheats toggles otherwise?

**Comment by MrAlaux** [comment]
About the freeze cheat, I've had it in Nugget Doom for a while now, and back when I was implementing it I quickly noticed that scrollers (both floors and walls) and swirling flats would still animate. I fixed those quirks, [here's the commit](https://github.com/MrAlaux/Nugget-Doom/commit/2fc313e59cd0464c56e396d4f5b09386716baa2b) if it's of any use.

**Comment by rfomin** [comment]
> I quickly noticed that scrollers (both floors and walls) and swirling flats would still animate. I fixed those quirks, [here's the commit](https://github.com/MrAlaux/Nugget-Doom/commit/2fc313e59cd0464c56e396d4f5b09386716baa2b) if it's of any use.

I like this quirk, although scrollers with things can look weird.

> Hm, on the other hand, these two just give a specific item. Aren't IDBEHOLD cheats toggles otherwise?

I've never used them, but it seems like they should give items. This is useful, I used to have a console command binding to give a medikit in a tight situation.

**Comment by fabiangreffrath** [comment]
Thanks for this! Though, now I really feel that we have too many cheat codes. Maybe it's time to reopen #800?

**Comment by rfomin** [comment]
> Thanks for this! Though, now I really feel that we have too many cheat codes. Maybe it's time to reopen #800?

I think we should keep all Boom cheats (documented here https://doomwiki.org/wiki/Source_port_cheat_codes). And all beta cheats. So I went through the tables and only `nuke` is a candidate for deletion.

Or we can remove the Boom cheats that are duplicated in the menu, such as `tran`, `smart` or `pitch`. Also I would like to move `showfps` to the menu.

Maybe we should document the cheats on a wiki like Crispy.

**Comment by rfomin** [comment]
> About the freeze cheat, I've had it in Nugget Doom for a while now, and back when I was implementing it I quickly noticed that scrollers (both floors and walls) and swirling flats would still animate. I fixed those quirks, [here's the commit](https://github.com/MrAlaux/Nugget-Doom/commit/2fc313e59cd0464c56e396d4f5b09386716baa2b) if it's of any use.

Although you're right, scroller interpolation in `frozen_mode` looks jerky. Fixed here https://github.com/fabiangreffrath/woof/pull/1001/commits/5c15d830bb66458b6f7f3b0730150b2dd12cb94f, thanks.

**Comment by fabiangreffrath** [comment]
My idea was to remove the hard-coded beta cheat and map them to the regular cheat codes with a dehacked patch in `autoload/betagrph.wad` directory.

**Comment by rfomin** [comment]
> My idea was to remove the hard-coded beta cheats and map them to the regular cheat codes with a dehacked patch in `autoload/betagrph.wad` directory.

OK, we can do that. Perhaps not in this PR?

**Comment by fabiangreffrath** [comment]
No, it's not that important.

**Review by fabiangreffrath — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
> I think we should keep all Boom cheats (documented here https://doomwiki.org/wiki/Source_port_cheat_codes). And all beta cheats. So I went through the tables and only `nuke` is a candidate for deletion.
> 
> Or we can remove the Boom cheats that are duplicated in the menu, such as `tran`, `smart` or `pitch`. Also I would like to move `showfps` to the menu.

So, we could still use the PR in #802 as a starting point. If we remap the beta cheats to regular cheats with a DEHACKED patch, we would still lose the ability to toggle auto-aim in beta mode, though (🤷).

> Maybe we should document the cheats on a wiki like Crispy.

Good idea, yes.

**Comment by rfomin** [comment]
> cheat removing

After some thought, I don't think we have that many cheats. What is the risk? Beta cheats only work in beta mode. The only advantage is less code, but we have a lot of room for refactoring elsewhere.

For consistency, I want to try to make `idbholdo` end with `0` like in Crispy.

**Comment by fabiangreffrath** [comment]
> After some thought, I don't think we have that many cheats. What is the risk? Beta cheats only work in beta mode. The only advantage is less code, but we have a lot of room for refactoring elsewhere.

Sure, but this loop is getting longer and longer:

https://github.com/fabiangreffrath/woof/blob/21b94c784630739df3c729aa47dd109d1658dca7/src/m_cheat.c#L1196-L1215

> For consistency, I want to try to make `idbholdo` end with `0` like in Crispy.

Yes, that'd be nice.

**Pull request merged**